circom/aes256.go

```diff
@@ -0,0 +1,311 @@
+			tmp2[g] = aes.api.Xor(aes.api.Xor(aes.api.Xor(a0Bits[g], a1gmc2Bits[g]), a2gmc3Bits[g]), a3Bits[g])
+			tmp3[g] = aes.api.Xor(aes.api.Xor(aes.api.Xor(a0Bits[g], a1Bits[g]), a2gmc2Bits[g]), a3gmc3Bits[g])
+			tmp4[g] = aes.api.Xor(aes.api.Xor(aes.api.Xor(a0gmc3Bits[g], a1Bits[g]), a2Bits[g]), a3gmc2Bits[g])
+		}
+		newState[c] = uapi.ByteValueOf(aes.api.FromBinary(tmp1...))
+		newState[4+c] = uapi.ByteValueOf(aes.api.FromBinary(tmp2...))
+		newState[8+c] = uapi.ByteValueOf(aes.api.FromBinary(tmp3...))
+		newState[12+c] = uapi.ByteValueOf(aes.api.FromBinary(tmp4...))
+	}
+	return newState
+}
+
+// required in mixcolumns
+func (aes *AES256) galoisMulConst(a uints.U8, idx int) uints.U8 {
+	p := uints.NewU8(0)
+	for counter := 0; counter < 8; counter++ {
+		if (idx & 1) != 0 {
+			p = aes.variableXor(p, a, 8)
+		}
+		idx = idx >> 1
+		if idx == 0 {
+			counter = 8
+			break
+		}
+
+		hiBitSet := aes.getBit(a, 8, 1)
+		a = aes.shiftLeft(a, 8, 1)
+		tmp := aes.variableXor(a, uints.NewU8(0x1B), 8)
+		sub := aes.api.Sub(tmp.Val, a.Val)
+		mul := aes.api.Mul(hiBitSet, sub)
+		a = uints.U8{Val: aes.api.Add(a.Val, mul)}
+	}
+	return p
+}
+
+// helper for galoisMul
+func (aes *AES256) getBit(a uints.U8, size, idx int) frontend.Variable {
+	bits := aes.api.ToBinary(a.Val, size)
+	return bits[len(bits)-idx]
+}
+
+// required for galoisMul
+func (aes *AES256) shiftLeft(a uints.U8, size, shift int) uints.U8 {
+	uapi, err := uints.New[uints.U32](aes.api)
+	if err != nil {
+		return uints.U8{}
+	}
+	bits := aes.api.ToBinary(a.Val, size)
+	x := make([]frontend.Variable, size)
+	for i := 0; i < size; i++ {
+		if i < shift {
+			x[i] = 0
+		} else {
+			x[i] = bits[i-shift]
+		}
+	}
+	bytes := aes.api.FromBinary(x...)
+	return uapi.ByteValueOf(bytes)
+}
+
+// shifts state matrix rows
+func (aes *AES256) shiftRows(state [16]uints.U8) [16]uints.U8 {
+	var newState [16]uints.U8
+	for i := 0; i < 4; i++ {
+		newState[i] = state[i] // 0, 1, 2, 3 == t0
+	}
+	for i := 0; i < 4; i++ {
+		newState[4+i] = state[4+((i+1)%4)] // 1, 2, 3, 0 == t1
+	}
+	for i := 0; i < 4; i++ {
+		newState[8+i] = state[8+((i+2)%4)] // 2, 3, 0, 1 == t2
+	}
+	for i := 0; i < 4; i++ {
+		newState[12+i] = state[12+((i+3)%4)] // 3, 0, 1, 2 == t3
+	}
+	return newState
+}
+
+// adds xor and shifts bytes in matrix to match next round representation requirements
+func (aes *AES256) addRoundKey(state [16]uints.U8, expandedKey [240]uints.U8, from int) [16]uints.U8 {
+	var newState [16]uints.U8
+	for i := 0; i < 4; i++ {
+		newState[i] = aes.variableXor(state[i], expandedKey[from+(4*i)], 8)
+		newState[4+i] = aes.variableXor(state[4+i], expandedKey[from+(4*i)+1], 8)
+		newState[8+i] = aes.variableXor(state[8+i], expandedKey[from+(4*i)+2], 8)
+		newState[12+i] = aes.variableXor(state[12+i], expandedKey[from+(4*i)+3], 8)
+	}
+	return newState
+}
+
+// different re-arrangement of variables
+func (aes *AES256) addRoundKey2(state [16]uints.U8, expandedKey [240]uints.U8, from int) [16]uints.U8 {
```

Comment:
Do you think we need bounds checks to ensure that array indices are always within bounds?

Reply:
Not too important. Because the input parameters have been limited. Because these are actually internal functions, users will not call these